graphql/src/values/coercion.rs

```diff
@@ -117,55 +132,69 @@ pub(crate) fn coerce_input_value<'a>(
     };
 
     Ok(Some(
-        coerce_value(&value, &def.value_type, resolver, variable_values).ok_or_else(|| {
+        coerce_value(value, &def.value_type, resolver, variable_values).map_err(|val| {
             QueryExecutionError::InvalidArgumentError(
                 def.position.clone(),
                 def.name.to_owned(),
-                value.clone(),
+                val,
             )
         })?,
     ))
 }
 
+/// On error, the `value` is returned as `Err(value)`.
 pub(crate) fn coerce_value<'a>(
-    value: &Value,
+    value: Value,
     ty: &Type,
     resolver: &impl Fn(&Name) -> Option<&'a TypeDefinition>,
     variable_values: &HashMap<q::Name, q::Value>,
-) -> Option<Value> {
+) -> Result<Value, Value> {
```

Comment:
This change is a really nice way to avoid cloning just so you can generate an error!